llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
+    calculateGOInfo(&Fn);
+    ++NumFunctions;
+  }
+#ifndef NDEBUG
+  for (auto &GI : GOIMap) {
+    LLVM_DEBUG(dbgs() << "\nGO Name:" << GI.first->getName() << "\n");
+    LLVM_DEBUG(GI.second.dump());
+  }
+#endif
 }
 
-// Loop through the contributions and add the initial digest of each global
-// object inside of the contributions to GOHash. Return true if GOhash is
-// changed.
-static bool updateDigestUseContributions(MD5 &GOHash,
-                                         const GOVec &Contributions,
-                                         GOInfoMap &GOIMap) {
+void ModuleHashGenerator::updateDigestUseContributions(
+    MD5 &GOHash, const GOVec &Contributions) {
   for (const GlobalObject *const G : Contributions) {
     GOHash.update(GOIMap[G].InitialDigest.Bytes);
   }
-  return !Contributions.empty();
+  // GOHash/Module is changed if Contributions is not empty.
+  Changed = Changed || !Contributions.empty();
 }
 
-// Loop through the dependences and add the final digest of each global object
-// inside of the dependences to GOHash.
-template <typename Function>
-static bool updateDigestUseDependencies(MD5 &GOHash, const GOVec &Dependencies,
-                                        GOStateMap &Visited, GOInfoMap &GOIMap,
-                                        Function AccumulateGODigest) {
-  bool Changed = false;
+auto ModuleHashGenerator::updateDigestUseDependencies(const GlobalObject *GO,
+                                                      MD5 &GOHash,
+                                                      unsigned GODepth,
+                                                      const GOVec &Dependencies,
+                                                      bool HasRepoDefinitionMD)
+    -> std::tuple<size_t, DigestType> {
+  auto LoopPoint = std::numeric_limits<size_t>::max();
   for (const GlobalObject *const G : Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    size_t GDepth;
+    DigestType GDigest;
+    std::tie(GDepth, GDigest) =
+        updateDigestUseDependenciesAndContributions(G, HasRepoDefinitionMD);
+    // A GO which loops back to itself doesn't count as a loop.
+    if (G != GO)
+      LoopPoint = std::min(LoopPoint, GDepth);
+    GOHash.update(GDigest.Bytes);
   }
-  return Changed;
+  // Encoded the 'End' in the GOhash after accumulating all GO's Dependencies.
+  GOHash.update(static_cast<char>(Tags::End));
+
+  DigestType Digest;
+  GOHash.final(Digest);
+
+  // If LoopPoint is bigger than GODepth, GO doesn't belong to any loops (except
+  // self-loop). Record GO's hash in memoized hashes.
```

Comment:
How about:
"Memoize GO's hash if GO doesn't belong to any loops (except self-loop).
Loop not present if LoopPoint is greater than current GO depth."

Reply:
Thanks.